src/common/imageio_libraw.c

```diff
@@ -53,32 +53,142 @@ static gboolean _supported_image(const gchar *filename)
+  if(!raw) return DT_IMAGEIO_FILE_CORRUPTED;
+
+  libraw_err = libraw_open_file(raw, filename);
+  if(libraw_err != LIBRAW_SUCCESS) goto error;
+
+  libraw_err = libraw_unpack(raw);
+  if(libraw_err != LIBRAW_SUCCESS) goto error;
+
+  // Bad method to detect if camera is fully supported by libraw.
+  // But seems to be the best available. libraw crx decoder can actually
+  // decode the raw data, but internal metadata like wb_coeffs, crops etc.
+  // are not populated into libraw structure.
+  if(raw->color.cam_mul[0] == 0.0)
+  {
+    libraw_close(raw);
+    return DT_IMAGEIO_FILE_CORRUPTED;
+  }
+
+  libraw_err = libraw_dcraw_process(raw);
+  if(libraw_err != LIBRAW_SUCCESS) goto error;
+
+  // Copy white level
+  img->raw_white_point = raw->color.maximum;
+
+  // Copy black level, specific for Canon as the regular black level
+  // info in libraw is set to 0. We need to look into makernotes!
+  img->raw_black_level_separate[0] = raw->makernotes.canon.ChannelBlackLevel[0];
+  img->raw_black_level_separate[1] = raw->makernotes.canon.ChannelBlackLevel[1];
+  img->raw_black_level_separate[2] = raw->makernotes.canon.ChannelBlackLevel[2];
+  img->raw_black_level_separate[3] = raw->makernotes.canon.ChannelBlackLevel[3];
+
+  // AsShot WB coeffs, caution: different ordering!
+  img->wb_coeffs[0] = raw->color.cam_mul[2];
+  img->wb_coeffs[1] = raw->color.cam_mul[1];
+  img->wb_coeffs[2] = raw->color.cam_mul[0];
+  img->wb_coeffs[3] = raw->color.cam_mul[3];
+
+  // Raw dimensions. This is the full sensor range.
+  img->width = raw->sizes.raw_width;
+  img->height = raw->sizes.raw_height;
+
+  // Apply crop parameters
+  img->crop_x = raw->sizes.raw_inset_crop.cleft;
+  img->crop_y = raw->sizes.raw_inset_crop.ctop;
+  img->crop_width = raw->sizes.raw_width - raw->sizes.raw_inset_crop.cwidth - raw->sizes.raw_inset_crop.cleft;
+  img->crop_height = raw->sizes.raw_height - raw->sizes.raw_inset_crop.cheight - raw->sizes.raw_inset_crop.ctop;
+
+  // We can reuse the libraw filters property, it's already well-handled in dt.
+  // It contains the BAYER filter pattern.
+  img->buf_dsc.filters = raw->idata.filters;
+
+  // For CR3, we only have BAYER data and a single channel
+  img->buf_dsc.channels = 1;
+
+  img->buf_dsc.datatype = TYPE_UINT16;
   img->buf_dsc.cst = iop_cs_RAW;
-  img->flags |= DT_IMAGE_4BAYER;
-  img->flags &= ~DT_IMAGE_LDR;
-  img->flags |= DT_IMAGE_RAW;
-  img->loader = LOADER_LIBRAW;
 
-  libraw_close(raw_data);
+  // Allocate and copy image from libraw buffer to dt
+  void *buf = dt_mipmap_cache_alloc(mbuf, img);
+  if(!buf)
+  {
+    err = DT_IMAGEIO_CACHE_FULL;
+    goto error;
+  }
+  dt_imageio_flip_buffers((char *)buf, (char *)raw->rawdata.raw_image, sizeof(uint16_t), raw->sizes.raw_width,
+                          raw->sizes.raw_height, raw->sizes.raw_width, raw->sizes.raw_height, raw->sizes.raw_pitch, ORIENTATION_NONE);
+
+
+  // Checks not really required for CR3 support, but it's taken from the old dt libraw integration.
```

Comment:
@TurboGit This is not really need for CR3, but I found this code in old dt-libraw implementation and the logic seems to be useful.